server/sonar-process-monitor/src/test/java/org/sonar/application/SchedulerImplTest.java

```diff
@@ -0,0 +1,384 @@
+  }
+
+  @Test
+  public void awaitTermination_blocks_until_all_processes_are_stopped() throws Exception {
+    Scheduler underTest = startAll();
+
+    Thread awaitingTermination = new Thread(() -> underTest.awaitTermination());
+    awaitingTermination.start();
+    assertThat(awaitingTermination.isAlive()).isTrue();
+
+    underTest.terminate();
+    // the thread is being stopped
+    awaitingTermination.join();
+    assertThat(awaitingTermination.isAlive()).isFalse();
+  }
+
+  @Test
+  public void restart_reloads_java_commands_and_restarts_all_processes() throws Exception {
+    Scheduler underTest = startAll();
+
+    processLauncher.waitForProcess(WEB_SERVER).askedForRestart = true;
+
+    // waiting for all processes to be stopped
+    boolean stopped = false;
+    while (!stopped) {
+      stopped = orderedStops.size() == 3;
+      Thread.sleep(1L);
+    }
+
+    // restarting
+    verify(appReloader, timeout(10_000)).reload(settings);
+    processLauncher.waitForProcessAlive(ELASTICSEARCH);
+    processLauncher.waitForProcessAlive(COMPUTE_ENGINE);
+    processLauncher.waitForProcessAlive(WEB_SERVER);
+
+    underTest.terminate();
+    // 3+3 processes have been stopped
+    assertThat(orderedStops).hasSize(6);
+    assertThat(processLauncher.waitForProcess(ELASTICSEARCH).isAlive()).isFalse();
+    assertThat(processLauncher.waitForProcess(COMPUTE_ENGINE).isAlive()).isFalse();
+    assertThat(processLauncher.waitForProcess(WEB_SERVER).isAlive()).isFalse();
+
+    // verify that awaitTermination() does not block
+    underTest.awaitTermination();
+  }
+
+  @Test
+  public void restart_stops_all_if_new_settings_are_not_allowed() throws Exception {
+    Scheduler underTest = startAll();
+    doThrow(new IllegalStateException("reload error")).when(appReloader).reload(settings);
+
+    processLauncher.waitForProcess(WEB_SERVER).askedForRestart = true;
+
+    // waiting for all processes to be stopped
+    processLauncher.waitForProcessDown(ELASTICSEARCH);
+    processLauncher.waitForProcessDown(COMPUTE_ENGINE);
+    processLauncher.waitForProcessDown(WEB_SERVER);
+
+    // verify that awaitTermination() does not block
+    underTest.awaitTermination();
+  }
+
+  private SchedulerImpl newScheduler() {
+    return new SchedulerImpl(settings, appReloader, javaCommandFactory, processLauncher, appState)
+      .setProcessWatcherDelayMs(1L);
+  }
+
+  private Scheduler startAll() throws InterruptedException {
+    enableAllProcesses();
+    SchedulerImpl scheduler = newScheduler();
+    scheduler.schedule();
+    processLauncher.waitForProcess(ELASTICSEARCH).operational = true;
+    processLauncher.waitForProcess(WEB_SERVER).operational = true;
+    processLauncher.waitForProcess(COMPUTE_ENGINE).operational = true;
+    return scheduler;
+  }
+
+  private void waitForAppStateOperational(ProcessId id) throws InterruptedException {
+    while (true) {
+      if (appState.isOperational(id)) {
+        return;
+      }
+      Thread.sleep(1L);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)